### ISSUE TITLE
Quiet boot option

### ISSUE THREAD

**Issue by miczyg1** [Dasharo/dasharo-issues#116]
**The problem you're addressing (if any)**
Currently Dasharo firmware prints strings on the screen indicating the menu prompt.

**Describe the solution you'd like**
Add a "Quiet boot" option to disable the strings during POST.

**Where is the value to a user, and who might that user be?**
These strings are not necessary and some people may not like it. There should be an option to disable them.

**Describe alternatives you've considered**
None

**Additional context**
None

**Comment by wessel-novacustom** [comment]
After almost two years, I haven't heard any user that is interested in this feature. I would close it without implementing.

**Comment by mkopec** [comment]
with quiet boot we could eliminate the flicker that happens when clearing the strings before handing off to OS :)

**Comment by wessel-novacustom** [comment]
This, plus skipping things like searching for boot options, so you will have a quiet-fast boot option. Maybe there are more elements that can be skipped during boot to realise a faster boot time.

**Comment by pietrushnic** [comment]
Any boot time optimization should start with measurement. What is the current firmware boot time, precisely?

**Comment by wessel-novacustom** [comment]
> While quiet boot is a simple modification, fast boot isn't. Also fast boot and quiet boot are two different things and should not be mixed.
> 
> > plus skipping things like searching for boot options
> 
> If we do that, everything will fall apart most likely. The drivers still need to initialize the storage to load the OS bootloader. It is not the boot option searching that takes most of the time, but I/O device enumeration and initialization from my experience. Searching for boot options is actually not that time consuming.
> 
> We could skip drivers that initialize everything else than SATA and NVMe, but then there is no way to enter setup during boot nor to have any input, if the USB and PS/2 drivers would not be loaded. I'm not aware of any good public reference for a fastboot option, thus I have no ideas how to approach it properly and not cause problems for device owners. Slimbootloader has fast boot but it essentially skips the graphics initialization, measured boot and verified boot (so basically strips security and output device initialization).

Thank you for the technical feedback on this.

Do you see any other option to reduce boot times?

**Comment by miczyg1** [comment]
> Do you see any other option to reduce boot times?

Like @pietrushnic said, measurements first. Then we can talk what we can improve. I/O device enumeration is what takes the most time from my experience, but it varies from platform to platform.

**Comment by pietrushnic** [comment]
IMHO, we should have reliable boot time measurements. I'm pretty sure we are doing it as part of testing, but do we have information on how much time each phase takes?

**Comment by Firminator** [comment]
@miczyg1 "prints strings on the screen indicating the menu prompt"
Are you refering to the 3 line menu on the upper left (DEL for Setup, F11 for Boot Menu, ENTER for Boot)? If not a screenshot would be helpful for context.

Other (proprietary) BIOS I have seen in the past have a "full screen boot logo" option. Dell devices, Lenovo laptops, and even the stock MSI Z690 I believe has it. This option sounds similar to what you like to achieve (just show the boot logo with no text).

Either way I agree measurement first before trying to optimize.

**Comment by miczyg1** [comment]
> Are you refering to the 3 line menu on the upper left (DEL for Setup, F11 for Boot Menu, ENTER for Boot)? If not a screenshot would be helpful for context.

Yes, these ones. There are also extra liens with firmware version and EC version printed on laptops too.

> Other (proprietary) BIOS I have seen in the past have a "full screen boot logo" option. Dell devices, Lenovo laptops, and even the stock MSI Z690 I believe has it. This option sounds similar to what you like to achieve (just show the boot logo with no text).

Full screen boot logo? No, that's not it. Boot logo does not have to be stretched to the whole screen. Quiet boot just means that nothing is printed on the screen except logo.

There is still open question what to do about warnings like vboot, battery, etc. Those should still remain IMO.

> IMHO, we should have reliable boot time measurements. I'm pretty sure we are doing it as part of testing, but do we have information on how much time each phase takes?

@pietrushnic coreboot has detailed timestamps at each initialization point. EDK2 seems to measure many events too based on the source code. However in Linux  we can only see:
`bootloader_launch_ns      bootloader_load_ns        exitbootservice_end_ns    exitbootservice_start_ns  firmware_start_ns ` in the sysfs of FPDT ACPI table. It looks like some kind of summarized info on the boot performance. But there should be more detailed records of the UEFI Payload execution. I just have to find it (and parse).

**Comment by miczyg1** [comment]
UEFI Shell has a command `dp` which displays the performance measurements. Sample output from MSI platform:

[boot_perf.log](https://github.com/user-attachments/files/18165879/boot_perf.log)

**Comment by pietrushnic** [comment]
This is a lovely dump we should get every time we release. The key question is what can be optimized. PS/2 timing looks crazy.

**Comment by Firminator** [comment]
Thanks @miczyg1 for figuring this out. After all those years I'm glad to see there is now **measurable** data on boot/initialization performance when disabling unused devices and controllers. I've been disabling the PS/2 and IDE/ATA and/or SATA controllers for years in my various builds in best faith that I can shave off a few milliseconds of boot time. Being able to measure this with DP is huge.
The Screenshot driver timings are off the chart. The screenshot functionality could be exposed in the Setup menu for the user to be able to enable-on-demand for when reporting an issue, but else keep it disabled by default.

@miczyg1 How do redirec the dp -v output to a USB flash drive?

**Comment by mkopec** [comment]
The screenshot delay is even more annoying, it's just stalling to display a status indicator in the screen corner :\ Can be removed or hidden behind a build option for our purposes IMO

**Comment by miczyg1** [comment]
> @miczyg1 How do redirec the dp -v output to a USB flash drive?

When entering the UEFI Shell, it will print available filesystems. You can redirect the output like on any other Linux. E.g.

```
Shell> FS0:
FS0> dp > boot_perf.log
```

**Comment by miczyg1** [comment]
> This is a lovely dump we should get every time we release. The key question is what can be optimized. PS/2 timing looks crazy.

We can't if we want to remove the UEFI Shell from builds. I recall we wanted to get rid of internal UEFI Shell because of Secure Boot. So we would have to think first about adding and option to boot UEFI Shell from external filesystem, subject to Secure Boot verification, if anything.

**Comment by pietrushnic** [comment]
Yes, we should carefully consider that. I hoped the Zarhus Team (@artur-rs cc) would update their UEFI Secure Boot tests. The best course of action is to create a Shell signed with robot-made keys, then those keys are provisioned, and we can do other tests like a Suite Prep procedure.

**Comment by miczyg1** [comment]
> Yes, we should carefully consider that. I hoped the Zarhus Team (@artur-rs cc) would update their UEFI Secure Boot tests. The best course of action is to create a Shell signed with robot-made keys, then those keys are provisioned, and we can do other tests like a Suite Prep procedure.

Tests may run UEFI SHell with Secure Boot disabled, no need to go to such lengths to sign the Shell binary. I think we have separate tests for testing the signing the verification.

**Comment by pietrushnic** [comment]
Yeah, it makes much more sense. I think just secure boot-related tests that rely on the uefi shell should be changed, but it is out of the scope of this issue.

**Comment by wessel-novacustom** [comment]
> > This is a lovely dump we should get every time we release. The key question is what can be optimized. PS/2 timing looks crazy.
> 
> We can't if we want to remove the UEFI Shell from builds. I recall we wanted to get rid of the internal UEFI Shell because of Secure Boot. So we would have to think first about adding an option to boot UEFI Shell from an external filesystem, subject to Secure Boot verification, if anything.
```

|￣￣￣￣￣￣￣￣￣￣￣￣￣￣| 
|    Make UEFI shell    | 
| disabled by default!  | 
|＿＿＿＿＿＿＿＿＿＿＿＿＿＿| 
     \          / 
      \ (•◡•)  /
       \      / 
        ——————
        |    | 
        |_   |_
```

**Comment by mkopec** [comment]
If we really need UEFI shell for some tests, we can simply put in on the ESP instead of including it in firmware. Then we could remove the builtin shell or build a stripped down version (MIN_SHELL)

**Comment by miczyg1** [comment]
After the analyzis of sources and the performance report pasted in the issue, the plan may look as follows:

1. Use PcdConInConnectOnDemand when fastboot enabled, to limit the console handling time and connect console input only when it is needed. OS bootloader may try to read keystrokes and it will trigger connecting the consoles, e.g. GRUB. Firmware will not set up console input before loading the OS bootloader. One may also request entering the FW setup from the OS itself:
- Windows: through [Windows Recovery Environment - Microsoft Support](https://support.microsoft.com/en-us/windows/windows-recovery-environment-0eb14733-6301-41cb-8d26-06a12b42770b)  (enter UEFI firmware settings)
- Linux: by executing command  systemctl reboot --firmware 
Once FW enters UI, it will connect all consoles and input will work.

2. Do not load/launch extra filesystem drivers on fastboot.
3. CrScreenshotDxe, get rid of status check in this line and 500ms stall: [CrScreenshotDxe/CrScreenshotDxe.c at master · LongSoft/CrScreenshotDxe](https://github.com/LongSoft/CrScreenshotDxe/blob/master/CrScreenshotDxe.c#L424) (fork most likely)
4. Only run VGA OptionROMs on fastboot.
5. Boot Timeout should be 0 on fastboot.

Of course adding 2 checkboxes in menu. I think the best place is Boot Mainetnance Manager, where the timeout lives. So:

Enable Fast Boot [] (default disabled)
Enable Quiet Boot [] (default disabled)

Enabling fastboot will case:
- boot timeout 0s (not updated in the menu, to avoid keeping the previous value in UEIF variables and occupying extra space), in short if fastboot then Timeout = 0 else ReadVar(Timeout)
- quiet boot (not updated in the menu, to avoid keeping the previous value in UEIF variables and occupying extra space), in short do not print prompts and progress bar if fastboot or quietboot

I think fastboot and quiet boot variable may be exposed to OS, so that anybody can disable it when needed.

Plus documenting new options, writing new tests specification.

BTW: we have an issue for fastboot: https://github.com/Dasharo/dasharo-issues/issues/115

**Comment by miczyg1** [comment]
PR: https://github.com/Dasharo/edk2/pull/204

**Comment by Firminator** [comment]
Will this fix (shaving off the 500ms from the Screenshot UEFI driver) be made available for the Z690?

**Comment by miczyg1** [comment]
> Will this fix (shaving off the 500ms from the Screenshot UEFI driver) be made available for the Z690?

It should naturally become available when a new release is out for the MSI boards (assuming PR will get merged).

**Comment by Firminator** [comment]
Sounds good.
On a sidenote how is it decided if the PR will be merged? I see that the Project page ( https://github.com/orgs/Dasharo/projects/8 ) doesn't have anything planned for the Z690 while the NovaCustom project page is full of Issues and PRs. Is the Z690 page even up to date?

**Comment by mkopec** [comment]
Implemented in https://github.com/Dasharo/edk2/pull/204
